### PR TITLE
Add my proposal

### PR DESCRIPTION
**Pull request by MizukiSonoko**
- Command should be enum. and separate command and object. I referred to @emilmelnikov 
- Domain should have ledgerId, Asset should not it.
- Add QueryFilter and Query has it.
- Add CommandOption for some command like Transfer

### CODE REVIEW

File: flatbuf/api.fbs [hunk 1]
```diff
@@ -83,9 +83,14 @@ table Signature {
   signature:string;
 }
 
+table QueryFilter {
```
Comment (l4l) [inline flatbuf/api.fbs]: Could you explain what does it intend to be/do?

File: flatbuf/api.fbs [hunk 2]
```diff
@@ -83,9 +83,14 @@ table Signature {
...
+  value: string;
+  condition: string;
+}
```
Comment (l4l) [inline flatbuf/api.fbs]: add new line between tables plz

File: flatbuf/api.fbs [hunk 3]
```diff
@@ -95,38 +100,23 @@ table ConsensusEvent {
...
-}
-
-table Remove {
-    object: Object(required);
-}
-
 table Transaction {
-  command: Command(required);
+  creatorPubKey: string(required);
+  command: string(required);
```
Comment (l4l) [inline flatbuf/api.fbs]: Why this is string?

File: flatbuf/api.fbs [hunk 4]
```diff
@@ -95,38 +100,23 @@ table ConsensusEvent {
...
-// Commands
-
-table Add {
-    object: Object(required);
-}
-
-table Transfer {
-    object: Object(required);
-    receiver: string(required);
+table CommandOption {
```
Comment (l4l) [inline flatbuf/api.fbs]: could you explain this as well?

**Comment by emilmelnikov** [comment]
Unfortunately, when I've suggested replacing commands with enums, I didn't think of command parameters (e.g. the _transfer_ command should have the _receiver_ parameter).
The solution with the `CommandOption` table doesn't look nice and complicates the ledger storage.

Maybe revert back to command tables/structs, but define them separately for each object type?

Example API sketch:

```
table Transaction {
  object: Object;
}

union Object { Currency, Peer }

table Currency {
  value: ulong;
  command: CurrencyCommand;
}

union CurrencyCommand { CurrencyAdd, CurrencyRemove, CurrencyTransfer }

struct CurrencyTransfer {
  receiver: string;
}
```

The downside of this approach is a lot of object-command tables.

**Comment by emilmelnikov** [comment]
The better option is to remove command/object nesting altogether:

```
table Transaction {
  command: Command;
}

union Command { CurrencyAdd, CurrencyRemove, CurrencyTransfer, PeerAdd, PeerRemove };

table CurrencyAdd {
  currency: Currency;
}

table CurrencyRemove {
  currency: Currency;
  // The alternative way is to list the "primary key" of the object:
  // name: string;
  // domain: Domain;
  // ledgerID: string;
}

table CurrencyTransfer {
  currency: Currency;
  // The "alternative way" also applies here.
  receiver: string;
}
```

In code, it will be handled with a one giant switch command, but one can hide it in some kind of factory:

```
class Command {
  virtual ~Command() {};
  // Common command interface...
}

class CurrencyAddCommand : Command {
  virtual ~CurrencyAddCommand() override {};
  // Implementation of the command interface...
}

class CurrencyRemoveCommand : Command {
  virtual ~CurrencyRemoveCommand() override {};
  // Implementation of the command interface...
}

class TransactionFactory {
  Command getCommand(const uint8_t* buffer);
}

Command TransactionFactory::getCommand(const uint8_t* buffer) {
  uint8_t* buf;
  auto transaction = GetTransaction(buffer);
  switch (transaction->command_type()) {
    case (Command_CurrencyAdd):
      // Handle CurrencyAdd
      return CurrencyAddCommand(params);
    case (Command_CurrencyRemove):
      // Handle CurrencyRemove
      return CurrencyRemoveCommand(params);
  }
}
```

**Comment by MizukiSonoko** [comment]
Ok thanks, I understanded it
I think code increase on large scale...

File: flatbuf/api.fbs (hunk 2 above)
Comment (MizukiSonoko) [inline flatbuf/api.fbs]: Ok thanks

File: flatbuf/api.fbs (hunk 1 above)
Comment (MizukiSonoko) [inline flatbuf/api.fbs]: QueryFilter is used by query.
For example
```
table QueryFilter {
  value: string = "my=asset-name"
  condition = "equal"
}
```
```
table QueryFilter {
  value: string = "value"
  condition = "not-equal"
}
```

**Comment by Warchant** [comment]
My proposal:
https://gist.github.com/Warchant/b7049e2f0653bd30c78f06b283c30bf0

Sorry, there are lots of files, probably some can be combined.
Also, don't worry about lots of includes:
> include automatically ensures each file is parsed just once, even when referred to more than once., -- Google

**Comment by takemiyamakoto** [comment]
Meeting tonight at 19:00 JST to discuss the data format some more.

**Comment by MizukiSonoko** [comment]
**Unnessecarry**

**Pull request closed without merging**